### PR TITLE
Create encrypt-what-is-necessary

### PR DESCRIPTION
**Pull request by gopaltayal**
### Description
Data protection through encryption is a crucial aspect of our security measures. However, the encryption process can be resource-intensive at multiple levels. Firstly, the amount of CPU required for encryption varies depending on the chosen algorithm, and more complex algorithms tend to demand higher computational power. Additionally, encryption can lead to increased storage requirements as it inflates the size of the data being stored because it typically contains additional metadata and padding, which is especially noticeable for smaller files. Furthermore, encryption is a repetitive task that needs to be performed each time data is fetched or updated. This repetitive nature can contribute to increased energy consumption, especially in high-throughput systems. Lastly, the most common approach involves encrypting the entire database or storage, without segregating sensitive and non-sensitive data. While this provides a high level of overall security, it also means that all data within the system incurs the energy cost associated with encryption, regardless of its sensitivity.

### Solution
1. Segregate data into sensitive and non-sensitive data and only encrypt the sensitive data.
2. Revise policy of what is considered sensitive and understand which environment can impact this sensitivity.
3. Choose the appropriate algorithm and the number of iterations in case of hashing.

### SCI Impact
SCI = (E * I) + M per R [Software Carbon Intensity Spec](https://grnsft.org/sci)
Only encrypting what needs to be encrypted will impact SCI as follows:
E: Reducing the amount of data that needs encryption will reduce the amount of energy needed to encrypt/decrypt, and store the data.

### Assumptions
We assume that a significant amount of data is not sensitive and can be stored unencrypted.

### Considerations
1. Client side encryption.
2. Separate storage for encrypted and non-encrypted data.
3. Hard to implement post application design.

### CODE REVIEW

**Comment by markus-gsf-seidl** [comment]
Hi!

Thanks for your PR!

I can understand the theoretical nature of this. If you do something that re-shapes the physical nature of things, it's going to cost physical resources (energy, heat, etc.).
On the other side, everyone is stating that encrypting is almost transparent because of special hardware and sometimes done even by default without the possibility to disable it (ex: https is mostly mandatory now). So I do not know if this is a viable way forward as we are facing a "encryption by default" - future.
Additionally I think that sensitivity does not have an large amount of impact in the decision if something is encrypted. As meta data can also use, for example, to personalize without the meta data itself looking sensitive. This makes things very hard.

---

We are currently in a restructure phase and pattern discussions have paused for the moment. If the restructure phase is over we can have a look at this.

**Comment by gopaltayal** [comment]
Hello!

Thank you for your feedback!

We had these specific points to support this PR:

1. There is hardware being utilised to encrypt data, even if it's a "special" hardware like a HSM, which directly leads to usage in energy. I believe the less it's utilised for this purpose, the less energy it would consume.

2. The pattern also talks about selecting the appropriate algorithm for encrypting and using a more complex one only if security demands it. This [research](https://eprints.soton.ac.uk/469889/1/Comparative_Analysis_of_Energy_Costs_of_Asymmetric_vs_Symmetric_Encryption_Based_Security_Applications.pdf) would support this point.

3. If not for sensitivity of the data, then why use encryption at all? I firmly believe that sensitivity is the main driving factor for encryption. By carefully selecting what is considered sensitive and only encrypting that data, I believe we can reduce the amount of energy the entire process would consume.

-----------------------------------------------------------------------------------------------------------------------

Please do let me know your thoughts on the same whenever you feel this can be looked at 🙂

**Comment by markus-gsf-seidl** [comment]
To your points: 
1. Is there some research which addition power requirements these do need? I know that, for example, power constrainted Apple devices are encrypted by default ( https://support.apple.com/guide/security/data-protection-overview-secf6276da8a/web#:~:text=Apple%20uses%20a%20technology%20called,a%20Mac%20with%20Apple%20silicon. ) even on the Apple Watch. This provides me a gut feeling that the power requirements are not as large as it might seem
2. Ok, I'm with you on that. Agreed.
3. There are many reasons: Metadata hiding, plausible deniability, privacy, etc. The internet is nowadays encrypted by default for various reasons. If you come from another side: If you only encrypt the necessary parts, everyone knows where the "good" and "interesting" data is and will focus on that.


From my point of view we could repurpose this pattern to something like: Choose the appropriate encryption level.

**Review by markus-gsf-seidl — CHANGES_REQUESTED** [review]
Looking at the pattern again, the changes we are talking are not very large. It would change the description to mentioning that encryption does (theoretical?) need time and energy and you should choose the appropriate encryption level for you application and performance requirements. If you can also go along with no encryption at all: good for the environment.
Can you rephrase it a bit more "freely" in the pattern, not requiring the separation of encryption and non-encrypted data?

File: docs/catalog/cloud/encrypt-what-is-necessary [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+### SCI Impact
+SCI = (E * I) + M per R [Software Carbon Intensity Spec](https://grnsft.org/sci)
+Only encrypting what needs to be encrypted will impact SCI as follows:
+E: Reducing the amount of data that needs encryption will reduce the amount of energy needed to encrypt/decrypt, and store the data.
+
+### Assumptions
+We assume that a significant amount of data is not sensitive and can be stored unencrypted.
+
+### Considerations
+1. Client side encryption.
```
Comment (markus-gsf-seidl) [inline docs/catalog/cloud/encrypt-what-is-necessary:18]: Tbh, client side encryption would not change the energy requirement needed, doesn't it? Depending on where specialised hardware is available, this could also worsen it.

**Comment by gopaltayal** [comment]
Hello!

Thank you for your feedback on this! I have some consolidated points around it here:

1. Regarding the additional power usage for the encryption process. I can refer [here](https://aws.amazon.com/cloudhsm/faqs/#:~:text=Yes.,can%20my%20application%20use%20CloudHSM%3F) and [here](https://aws.amazon.com/kms/faqs/) on how various encryption services/dedicated modules for encryption work even within the cloud.

2. I agree with the part where people (I assume you mean people who want to cause security breaches) will focus on where the good parts are but isn't that the use case of encryption? To hide these parts. By choosing what we encrypt and what we don't encrypt we are not proposing that everything is left unencrypted and insecure. We are proposing to reduce the data that needs to be encrypted and not to encrypt full volumes/drives or storage spaces and instead only the data that needs this encryption.

3. I will rephrase the pattern more "freely" 🙂 and add the part of segregation of data to the considerations, which anyone implementing can decide to use/not use.

**Comment by markus-gsf-seidl** [comment]
I like it! Thanks @gopaltayal 

There is one technical thing: The markdown file is missing the header from the template, and maybe some other structural things. I feel it needs comparison with the original pattern template and fill in all the missing. @russelltrow Maybe someone with write rights can change this PR and fill that in? Or shall we copy it?

**Comment by gopaltayal** [comment]
@markus-gsf-seidl could you please let me know more about the structural changes? I am currently using the [Template](https://github.com/Green-Software-Foundation/patterns/blob/dev/TEMPLATE.md) provided in the repository to structure it, but I see that in the [Readme](https://github.com/Green-Software-Foundation/patterns/blob/dev/README.md) that the components are provided differently. Which one would be the correct one to follow? I could make the changes accordingly 🙂

**Comment by russelltrow** [comment]
> @markus-gsf-seidl could you please let me know more about the structural changes? I am currently using the [Template](https://github.com/Green-Software-Foundation/patterns/blob/dev/TEMPLATE.md) provided in the repository to structure it, but I see that in the [Readme](https://github.com/Green-Software-Foundation/patterns/blob/dev/README.md) that the components are provided differently. Which one would be the correct one to follow? I could make the changes accordingly 🙂

Hi @gopaltayal thank you so much for submitting the pattern. You are using the correct Template, there are just a few header fields that we ask you to add please. These are clearer on the [raw template view](https://github.com/Green-Software-Foundation/patterns/blob/dev/TEMPLATE.md?plain=1).

I've just commented on your commit with a suggested set of tags.

File: docs/catalog/cloud/encrypt-what-is-necessary [hunk 2]
```diff
@@ -0,0 +1,23 @@
+### Description
```
Comment (russelltrow) [inline docs/catalog/cloud/encrypt-what-is-necessary:1]: Suggested header tags.

---
version: 1.0
submitted_by: @gopaltayal 
published_date: TBD
category: cloud
description: Data protection through encryption is a crucial aspect of our security measures. However, the encryption process can be resource-intensive at multiple levels.
tags:
- cloud
- security
---

**Comment by gopaltayal** [comment]
@russelltrow Thanks for letting me know! I have added the header fields, please do have a look and let me know if it looks okay to you.

File: docs/catalog/cloud/encrypt-what-is-necessary [hunk 3]
```diff
@@ -0,0 +1,39 @@
+---
+version: 1.0
+submitted_by: @amexboy @gopaltayal @samudurand @shainu-x
```
Comment (markus-gsf-seidl) [inline docs/catalog/cloud/encrypt-what-is-necessary:3]: I'm not a 100% sure which code evaluates this (if any), but all other patterns do not have the @ and separate by commas, instead of spaces (IIRC).

File: docs/catalog/cloud/encrypt-what-is-necessary [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+published_date: TBD
+category: cloud
+description: Data protection through encryption is a crucial aspect of our security measures. However, the encryption process can be resource-intensive at multiple levels.
+tags:
```
Comment (markus-gsf-seidl) [inline docs/catalog/cloud/encrypt-what-is-necessary:7]: Historyically (?) this should list "- size:medium"

File: docs/catalog/cloud/encrypt-what-is-necessary [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+## SCI Impact
+SCI = (E * I) + M per R [Software Carbon Intensity Spec](https://grnsft.org/sci)
+Only encrypting what needs to be encrypted will impact SCI as follows:
+E: Reducing the amount of data that needs encryption will reduce the amount of energy needed to encrypt/decrypt, and store the data.
+
+## Assumptions
+We assume that a significant amount of data is not sensitive and can be stored unencrypted.
+
+## Considerations
+1. Segregate data into sensitive and non-sensitive data and only encrypt the sensitive data.
```
Comment (markus-gsf-seidl) [inline docs/catalog/cloud/encrypt-what-is-necessary:34]: This is a unnumbered list.

File: docs/catalog/cloud/encrypt-what-is-necessary [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+# Encrypt What Is Necessary
+
+## Description
+Data protection through encryption is a crucial aspect of our security measures. However, the encryption process can be resource-intensive at multiple levels.
+Firstly, the amount of CPU required for encryption varies depending on the chosen algorithm, and more complex algorithms tend to demand higher computational power.
+Additionally, encryption can lead to increased storage requirements as it inflates the size of the data being stored because it typically contains additional metadata and padding, which is especially noticeable for smaller files.
+Furthermore, encryption is a repetitive task that needs to be performed each time data is fetched or updated. This repetitive nature can contribute to increased energy consumption, especially in high-throughput systems.
+
+## Solution
+1. Carefully consider the sensitivity of the data that the application processes to understand if encryption is needed or not.
```
Comment (markus-gsf-seidl) [inline docs/catalog/cloud/encrypt-what-is-necessary:21]: This is a unnumbered list.

File: docs/catalog/cloud/encrypt-what-is-necessary [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+2. Revise policy of what is considered sensitive and understand which environment can impact this sensitivity.
+3. Choose the appropriate algorithm and the number of iterations in case of hashing based on your application's performance and security needs.
+
+## SCI Impact
+SCI = (E * I) + M per R [Software Carbon Intensity Spec](https://grnsft.org/sci)
```
Comment (markus-gsf-seidl) [inline docs/catalog/cloud/encrypt-what-is-necessary:26]: It's `SCI ... ` (with the backquotes)

File: docs/catalog/cloud/encrypt-what-is-necessary [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+## Assumptions
+We assume that a significant amount of data is not sensitive and can be stored unencrypted.
+
+## Considerations
+1. Segregate data into sensitive and non-sensitive data and only encrypt the sensitive data.
+2. Separate storage for encrypted and non-encrypted data.
+3. Hard to implement post application design.
+
+## References
+- [Comparative Analysis of Energy Costs of Asymmetric vs Symmetric Encryption-Based Security Applications](https://eprints.soton.ac.uk/469889/1/Comparative_Analysis_of_Energy_Costs_of_Asymmetric_vs_Symmetric_Encryption_Based_Security_Applications.pdf)
```
Comment (markus-gsf-seidl) [inline docs/catalog/cloud/encrypt-what-is-necessary:39]: Since this is a scientific article, DOI links should be used: https://doi.org/10.1109/ACCESS.2022.3192970 . These links are then maintained by the DOI foundation.

**Review by markus-gsf-seidl — CHANGES_REQUESTED** [review]
Minor nitpicks added for style points.
I lack the rights to edit the PR myself.

**Comment by russelltrow** [comment]
@markus-gsf-seidl would you mind reviewing the changes please

**Comment by markus-gsf-seidl** [comment]
LGTM

**Review by markus-gsf-seidl — APPROVED** [review]
